[MRG] MLP - add class weights support

dorcoh opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #9113 
(should close stalled pull request #10588)

#### What does this implement/fix? Explain your changes.

- [x] Add sample_weight
  - [x] Add support to multiply sample weight sample wise on loss functions:
    - [x] log_loss
    - [x] binary_log_loss
    - [x] squared_loss
  - [x] Add sample_weight parameter to fit and partial_fit
    - [x] Propagate to relevant functions
    - [x] Validate sample_weight size on fit
  - [x] Add tests
    - [x] Test directly loss functions sample_weight invariance 
    - [x] Test fit validation of sample_weight size

- [x]  Add class_weight support
   - [x] Add support on fit and partial fit
   - [x] Add test for behaviour correctness 

- [ ] Test with `utils.estimator_checks.check_class_weight_classifiers`: Currently (Nov 18) have issues passing this test, however succeeded when using one neuron MLP
#### Any other comments?

I started by adding sample_weight parameter support, next should add class_weight.
Please share your comments

Thanks

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

dorcoh committed:
added sample weight support to all mlp loss functions

dorcoh committed:
added sample weight param to fit and partial fit; validated its size

dorcoh committed:
added tests for sample weight invariance on loss functions; sample weight size validation

dorcoh committed:
fixed flakes errors

dorcoh committed:
fixed flakes errors

dorcoh commented:
@jnothman, can you explain what do you mean by testing weighted back-propagation? 

Because you mentioned on the issue thread (#9113) the batch sampling (and adaptive learning rate) I assume you refer to this part on _fit_stochastic, where the outer loop control the epochs and inner loop the batches:

```python
for it in range(self.max_iter):
    X, y = shuffle(X, y, random_state=self._random_state)
    accumulated_loss = 0.0
    for batch_slice in gen_batches(n_samples, batch_size):
        activations[0] = X[batch_slice]
        batch_loss, coef_grads, intercept_grads = self._backprop(
            X[batch_slice], y[batch_slice], activations, deltas,
            coef_grads, intercept_grads,
            sample_weight[batch_slice])
        accumulated_loss += batch_loss * (batch_slice.stop -
                                          batch_slice.start)

        # update weights
        grads = coef_grads + intercept_grads
        self._optimizer.update_params(grads)

    self.n_iter_ += 1
    self.loss_ = accumulated_loss / X.shape[0]

    self.t_ += n_samples
    self.loss_curve_.append(self.loss_)
```

what should I test here then? I guess the weights correspond to (batch_slice.stop - batch_slice.start)

Thanks

jnothman reviewed (commented):
Hmm... Maybe I was just confused. But we do need some kind of end-to-end test that sample_weight has some effect, and ideally the correct effect (although perhaps that can simply be tested through class_weight when implemented).

jnothman commented:
No, let's not worry about weighted sampling I think, though it should be
noted somewhere that the weights do not affect the sampling.

On 1 August 2018 at 16:04, Dor Cohen <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman>, can you explain what do you mean
> by testing weighted back-propagation?
>
> Because you mentioned on the issue thread (#9113
> <https://github.com/scikit-learn/scikit-learn/issues/9113>) the batch
> sampling (and adaptive learning rate) I assume you refer to this part on
> _fit_stochastic, where the outer loop control the epochs and inner loop the
> batches:
>
> for it in range(self.max_iter):
>     X, y = shuffle(X, y, random_state=self._random_state)
>     accumulated_loss = 0.0
>     for batch_slice in gen_batches(n_samples, batch_size):
>         activations[0] = X[batch_slice]
>         batch_loss, coef_grads, intercept_grads = self._backprop(
>             X[batch_slice], y[batch_slice], activations, deltas,
>             coef_grads, intercept_grads,
>             sample_weight[batch_slice])
>         accumulated_loss += batch_loss * (batch_slice.stop -
>                                           batch_slice.start)
>
>         # update weights
>         grads = coef_grads + intercept_grads
>         self._optimizer.update_params(grads)
>
>     self.n_iter_ += 1
>     self.loss_ = accumulated_loss / X.shape[0]
>
>     self.t_ += n_samples
>     self.loss_curve_.append(self.loss_)
>
> what should I test here then? I guess the weights correspond to
> (batch_slice.stop - batch_slice.start)
>
> Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/11723#issuecomment-409459453>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60EoK6w9sqSqgHCf4D7VG7Z5wl86ks5uMUTygaJpZM4VpH5j>
> .
>

dorcoh commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -25,7 +25,10 @@
-
+from sklearn.neural_network._base import (log_loss, binary_log_loss,
+                                          squared_loss)
+from sklearn.metrics.tests.test_common import check_sample_weight_invariance
```
Will copy it to utils.testing then.

dorcoh commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -633,3 +636,61 @@ def test_n_iter_no_change_inf():
+    y = y_digits_binary[:n_samples]
+    # size differ
+    sample_weight = np.ones(n_samples-1, dtype=np.float64, order='C')
+    message = ('Shapes of X and sample_weight do not match.')
```
There's a whole module for testing compute_class/sample weight - test_class_weight under utils.tests, but it doesn't cover testing through calling fit. I can extend this current test to perform also similar check to this one:

```python
def test_gnb_sample_weight():
    """Test whether sample weights are properly used in GNB. """
    # Sample weights all being 1 should not change results
    sw = np.ones(6)
    clf = GaussianNB().fit(X, y)
    clf_sw = GaussianNB().fit(X, y, sw)

    assert_array_almost_equal(clf.theta_, clf_sw.theta_)
    assert_array_almost_equal(clf.sigma_, clf_sw.sigma_)

    # Fitting twice with half sample-weights should result
    # in same result as fitting once with full weights
    sw = rng.rand(y.shape[0])
    clf1 = GaussianNB().fit(X, y, sample_weight=sw)
    clf2 = GaussianNB().partial_fit(X, y, classes=[1, 2], sample_weight=sw / 2)
    clf2.partial_fit(X, y, sample_weight=sw / 2)

    assert_array_almost_equal(clf1.theta_, clf2.theta_)
    assert_array_almost_equal(clf1.sigma_, clf2.sigma_)

    # Check that duplicate entries and correspondingly increased sample
    # weights yield the same result
    ind = rng.randint(0, X.shape[0], 20)
    sample_weight = np.bincount(ind, minlength=X.shape[0])

    clf_dupl = GaussianNB().fit(X[ind], y[ind])
    clf_sw = GaussianNB().fit(X, y, sample_weight)

    assert_array_almost_equal(clf_dupl.theta_, clf_sw.theta_)
    assert_array_almost_equal(clf_dupl.sigma_, clf_sw.sigma_)
```

dorcoh commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -633,3 +636,61 @@ def test_n_iter_no_change_inf():
+    assert_raise_message(ValueError, message, clf.fit, X, y, sample_weight)
+    # partial fit
+    clf = MLPClassifier()
+    for i in range(42):
```
Just an arbitrary number because it is partial fit that's tested.

dorcoh committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into mlp_class_weights

dorcoh committed:
fixed trailing space

dorcoh committed:
added class_weight support to mlp fit and partial_fit

dorcoh committed:
modified sample weight test; added test for class weight

dorcoh commented:
Hi, so I changed a bit the sample weight test through fit,

Then added class weight support to fit and partial_fit, and lastly I added another test - for checking class_weight, seems like it's working fine, be glad to get your comments though.

Thanks

dorcoh commented:
btw, I got this error on test_docstring_parameters, apparently after adding class_weight to MLP fit functions. Do you know what does it mean?
```
E           AssertionError: Docstring Error: 
E           sklearn.neural_network.multilayer_perceptron.MLPRegressor.fit class_weight != sample_weight
E           sklearn.neural_network.multilayer_perceptron.MLPRegressor.fit sample_weight != class_weight
```

dorcoh committed:
changed fit docstring

dorcoh committed:
fixed flakes errors

dorcoh commented:
Fixed the last errors,

now another weird bug fails travis:
```
CondaHTTPError: HTTP 504 GATEWAY TIME-OUT for url <https://repo.anaconda.com/pkgs/main/linux-64/coverage-4.5.1-py36h14c3975_0.tar.bz2>
Elapsed: 00:10.011294
CF-RAY: 443e8d23fae4472e-EWR
An HTTP error occurred when trying to retrieve this URL.
```

dorcoh committed:
modified fit/partial test to include class_weight param

dorcoh committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into mlp_class_weights

dorcoh commented:
Last errors travis errors are for flakes, the class_weight docstring is too long but that's how it appears on other modules, let me know if to change it or break those lines.

jnothman commented:
Please ping after we have released 0.20 (or at least the RC

Thanks for some good work

jnothman commented:
Is there any reason this is still labelled WIP? Is there work remaining that makes it a problem if this was merged?

dorcoh commented:
Hi @jnothman , it's actually labelled MRG. 
Anyway it is ok by me, I wait for your reviews, thanks.

jnothman commented:
Oh. I got confused switching between GitHub and GMail...

jnothman commented on sklearn/neural_network/_base.py:
```diff
@@ -175,7 +177,18 @@ def inplace_relu_derivative(Z, delta):
 
 
-def squared_loss(y_true, y_pred):
+def _finalize_loss(y_true, y_prob, sample_weight, sample_score):
```
I think np.average does the same as this function. What am I missing?

dorcoh replied:
I see, you are right, `np.average` also validates the sizes for us.

jnothman commented on sklearn/neural_network/multilayer_perceptron.py:
```diff
@@ -611,11 +641,24 @@ def fit(self, X, y):
+        sample_weight : array-like, shape (n_samples,), optional (default=None)
+            Weights applied to individual samples (1. for unweighted).
+
+        class_weight : dict or 'balanced', optional
```
Parameters to fit should be sample-wise. class_weight should be an estimator `__init__` parameter

dorcoh replied:
I see, you are right, `np.average` also validates the sizes for us.

jnothman commented on sklearn/neural_network/multilayer_perceptron.py:
```diff
@@ -324,10 +330,18 @@ def _fit(self, X, y, incremental=False):
         X, y = self._validate_input(X, y, incremental)
         n_samples, n_features = X.shape
 
+        # validation: if None set unit weights.
```
If setting unit weights, why do loss functions handle the sample_weight is None case? However, I'd rather leave this None if we can, unless we show that this results in negligible efficiency reduction.

dorcoh replied:
Where should it be handled then?

jnothman reviewed (commented):
I'm yet to review tests. You have flake8 failure

jnothman commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -633,3 +638,213 @@ def test_n_iter_no_change_inf():
+
+def test_mlp_class_weight():
+    """ test class weights, should validate also if
+    sample weight effect is correct. aided by ridge module"""
```
Do you mean that you copied this from ridge? I'm not sure what the point of merging such a comment into the code base is

dorcoh replied:
No I haven't copied it, just helped by it, you can see I test other stuff. I will delete this comment, it doesn't add any information..

jnothman commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -633,3 +638,213 @@ def test_n_iter_no_change_inf():
+                             classes=np.unique(y),
+                             sample_weight=sample_weight)
+
+    # None sample weight or all being 1 predicts same output
```
This should already be tested in sklearn/utils/estimator_checks.py in `check_sample_weights_invariance`. You don't need it here

dorcoh replied:
Ok, I will explore this module.

jnothman commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -633,3 +638,213 @@ def test_n_iter_no_change_inf():
+        assert_array_equal(predicted, predicted_sw)
+
+
+def test_mlp_class_weight():
```
There are also generic tests for class_weight in estimator_checks.py

What's missing from estimator_checks is a test that sample_weight has any (or the right) effect.

dorcoh replied:
Will check it out.

Anyway about the sample_weight effect I made the hyper plane test, where it moves by giving low weights to one class.

And another test is the following:
```
    # data for balance check
    X_bal = np.array([[-1.0, -1.0], [-1.0, 0], [-.8, -1.0], [1.0, 1.0]])
    y_bal = [1, 1, -1, -1]
    # class_weight = 'balanced', and class_weight = None should return
    # same values when y has equal number of all labels
    clf.fit(X_bal, y_bal, class_weight=None)
    prob = clf.predict_proba(test_sample)

    clf.fit(X_bal, y_bal, class_weight='balanced')
    prob_b = clf.predict_proba(test_sample)

    assert_equal(len(clf.classes_), 2)
    assert_array_almost_equal(prob, prob_b)

```

jnothman commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -633,3 +638,213 @@ def test_n_iter_no_change_inf():
+
+
+@ignore_warnings
+def check_sample_weight_invariance(name, metric, y1, y2):
```
It would be nice if we could find a way to share rather than copy this code. Putting it in utils/testing.py might be the right thing?

dorcoh replied:
I guess so, will check if it works out.

dorcoh committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into mlp_class_weights

dorcoh committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into mlp_class_weights

dorcoh committed:
removed _finalize_loss helper; used np.average instead

dorcoh committed:
added class_weight to MLPClassifier constructor

dorcoh committed:
removed redundant functions and tests

dorcoh committed:
moved check_sample_weight_invariance to utils.testing

dorcoh committed:
skipped class weight for MLP in check_class_weight_classifiers

dorcoh commented:
Ok @jnothman  so I handled all of your comments,

Though there is one thing, the class weight test inside `estimator_checks` fails for MLP (see output below) so I skipped it currently. I tried to debug it but I'm not sure if it should even work (is linear classifier assumed there?). In any case I still test class_weight behavior through test_mlp.

Let me know how you want to handle it

```python
___________________________________________________________ test_non_meta_estimators[MLPClassifier-MLPClassifier-check_class_weight_classifiers] ___________________________________________________________

name = 'MLPClassifier', Estimator = <class 'sklearn.neural_network.multilayer_perceptron.MLPClassifier'>, check = <function check_class_weight_classifiers at 0x7f3ce6ae6ae8>

    @pytest.mark.parametrize(
            "name, Estimator, check",
            _generate_checks_per_estimator(_yield_all_checks,
                                           _tested_non_meta_estimators()),
            ids=_rename_partial
    )
    def test_non_meta_estimators(name, Estimator, check):
        # Common tests for non-meta estimators
        with ignore_warnings(category=(DeprecationWarning, ConvergenceWarning,
                                       UserWarning, FutureWarning)):
            estimator = Estimator()
            set_checking_parameters(estimator)
>           check(name, estimator)

tests/test_common.py:100: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
utils/testing.py:352: in wrapper
    return fn(*args, **kwargs)
utils/estimator_checks.py:1887: in check_class_weight_classifiers
    assert_greater(np.mean(y_pred == 0), 0.87)
/usr/lib/python3.5/unittest/case.py:1211: in assertGreater
    self.fail(self._formatMessage(msg, standardMsg))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sklearn.utils._unittest_backport.TestCase testMethod=__init__>, msg = '0.34 not greater than 0.87'

    def fail(self, msg=None):
        """Fail immediately, with the given message."""
>       raise self.failureException(msg)
E       AssertionError: 0.34 not greater than 0.87

/usr/lib/python3.5/unittest/case.py:665: AssertionError
```

dorcoh committed:
added docstring to check_sample_weight_invariance

dorcoh commented:
Last fail is for adding new MLP params to docs

dorcoh commented:
Hi @jnothman  , can we please take this forward?

jnothman commented:
Hi Dor, please be patient. The 0.20 release has been too big and too slow. 

Regarding the common test failure, I would intuitively expect MLP to pass that test, if the fit converges. Does it?

jnothman commented:
All `check_class_weight_classifiers` seems to be implying is that if there are a mix of points in three classes in the training data, then class 0 should be predicted most of the time if and only if it has non-negligible weight. It doesn't rely on linear models (and MLP should be able to find a linear fit in any case).

It might be a more sensible test if the classes are even more mixed, e.g. rather than `make_blobs`, draw all points from the same distribution.

"AssertionError: 0.34 not greater than 0.87" does seem to be problematic. It implies that the classifier is not biased to class 0 despite the weights.

jnothman commented:
It seems from a bit of testing that MLP is just far too unstable. The model
is completely different with each fit, but there is also no clear trend
resulting from skewed class weight.

jnothman commented:
Can we think of a more well conditioned learning problem where MLP will consistently respond to class_weight?

Yes, the doctests in `neural_networks_supervised.rst` need updating to mention class_weight=None.

dorcoh commented:
Sorry for the late response, about MLP being unstable I noticed it too when I've tested it.
will probably have more time this week to go over it and I'll update, thanks @jnothman

dorcoh committed:
Resolved conflicts by keeping master (new imports)

dorcoh commented:
Hi @jnothman , I made some plots using the classifier comparison example.

The two last rows depict the blobs dataset with two or three centers (the kind we use at `estimator_checks.py`), and indeed they doesn't seem to converge. 

First row is circles dataset, and then linearly separable, which do converge and I think we can use them. For example move the boundary of linearly separable dataset using the class weights, we can simulate it and verify the expected score without class weight (e.g., 10 positive vs 100 negative examples).

Actually I already made a similar test with toy data in `test_mlp.test_mlp_class_weight`

Let me know what you think.

![image](https://user-images.githubusercontent.com/7069817/46766430-a2c47e00-ccea-11e8-8b36-a837386a6f89.png)

jnothman commented:
I still don't understand why fitting noise with strong weights on one class
or another shouldn't work... The loss should be very different for the
weighted samples. Is the per-sample loss being normalised somewhere that
might be messing this up?

I don't see the linear test as being applicable to all classifiers either.

Does test_mlp_class_weight pass with most/all random_states?

dorcoh commented:
Hi, I made the plot with the stable version, so my modifications shouldn't effect it.

Maybe we can still work with blobs, but change the tolerance for convergence, I will try it later today.

About test_mlp_class_weight, I have tried only with one random state.

dorcoh committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into mlp_class_weights

jnothman commented:
We could change n_iter_no_change and/or tol for the checks. Another thing
we could do is reduce the model size so that it's just a logistic
regression or closer to it.

Do you mean that changing the number of samples still results in a fit
where the predicted class is random?

dorcoh commented:
Hi, I will try your suggestions

What do you mean the predicted class is random?

I tried to simulate the way the `check_class_weight_classifiers` should work on MLP class weight parameter, but without using it, this test scales one class weight by 0.001 and validates the probability for predicting the other class is over some ratio. So I generated many positive vs few negative examples, and tried to see (with the stable version) if the model converges. I made this experiment once with blobs, and once with gaussians.

In any case I will check it again and update

christinebuckler commented:
Just checking in on this issue as I am also interested in this functionality. When is 0.20 targeted to be released? Do you think this will be included?

jnothman commented:
0.20 has been released. This might be in 0.21

dorcoh committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into mlp_class_weights

dorcoh committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into mlp_class_weights

dorcoh committed:
added _IS_32BIT to testing.py

dorcoh commented:
Hi, checked it again, but this time tried to keep it simple.

I checked directly `check_class_weight_classifiers` with pytest, took your advice and changed the std to be 0 in blobs (from std=20), the tests failed again for the same reason (for 2 and for 3 centers). Though I plotted the losses for each of the tries, notice the scales are different for each plot, and it also stops sometimes below 1000 iterations (due to early stopping), class weight values are the same used on test.

# 2 centers
## 20 std
### class weight
![image](https://user-images.githubusercontent.com/7069817/48514022-0bf95d00-e866-11e8-947a-03cc5c4579c3.png)

### equal weights
![image](https://user-images.githubusercontent.com/7069817/48514099-2c291c00-e866-11e8-8034-a1cf0f9dfb07.png)

## 0 std
### class weights
![image](https://user-images.githubusercontent.com/7069817/48514383-ddc84d00-e866-11e8-84a8-55220da59643.png)

### equal weights
![image](https://user-images.githubusercontent.com/7069817/48514417-ef115980-e866-11e8-8840-6dab0f251fb1.png)

# 3 centers
## 20 std
### class weights
![image](https://user-images.githubusercontent.com/7069817/48514438-fdf80c00-e866-11e8-9825-d753abac89fd.png)

### equal weights
![image](https://user-images.githubusercontent.com/7069817/48514466-09e3ce00-e867-11e8-9c1c-268fde67d6b8.png)

## 0 std
### class weights
![image](https://user-images.githubusercontent.com/7069817/48514485-12d49f80-e867-11e8-8c37-993938f56912.png)

### equal weights
![image](https://user-images.githubusercontent.com/7069817/48514495-19631700-e867-11e8-8aba-d6f802f9342f.png)

Equal weights were for sanity check (I have actually set them to test the parameter) and indeed there are differences, moreover it looks like the zero std makes the optimization 'easier', but original test failed. I'm still not sure if my implementation is broken or if it's normal behavior, can't find anything yet.

Maybe we can design even more simple test, to take into account each sample contribution to the overall loss while training, if there's no difference then clearly something is broken. Let me know what you think, thanks!

jnothman commented:
So obviously the weighting has some effect. I'm on the move so I've not yet
had time to focus on what exactly could be happening here.

Testing the effect of a single sample may be appropriate, except that I'm
concerned that the MLP is overfitting and in general we're running into a
problem of having too much capacity to learn a reasonable model.

Have you tried running the test on a single-layer single latent element MLP
(i.e. a logistic unit)?

dorcoh commented:
Hi, about 1 neuron - it works for 2 centers, np.mean(y_pred == 0) is actually 1. For 3 centers it works too (accuracy 1) but for the wrong class, though I think it's something else. 
The method in [utils.class_weight.compute_sample_weight](https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_sample_weight.html) for handling the class weight parameter is not clear to me, specifically for multi-class , I had to change the test as following:

```python
#class_weight = {0: 1000, 1: 0.0001, 2: 0.0001}
class_weight = [{0:1, 1:1000}, {0:1, 1: 0.0001}, {0:1, 1: 0.0001}]
```
Is that correct ?

jnothman commented:
Surely that's for multilabel, not multiclass. I've not had enough time for
this lately to get into it deep. If someone else is able to help us ensure
this is all working correctly, that would be great...

jnothman commented:
@thomasjpfan, I appreciate you're working on a lot of things... but I wonder if you could help getting to the bottom of this: this much-requested enhancement has been implemented, but our common test for class weight support seems to fail the implementation. Questions include: is weighted MLP actually working properly? and can our common test be improved?

If not @thomasjpfan, would someone else like to drill down into this?

thomasjpfan commented:
@jnothman Thank you for the ping! I will look into this issue/PR.

thomasjpfan commented on sklearn/neural_network/multilayer_perceptron.py:
```diff
@@ -509,7 +537,8 @@ def _fit_stochastic(self, X, y, activations, deltas, coef_grads,
                         X[batch_slice], y[batch_slice], activations, deltas,
-                        coef_grads, intercept_grads)
+                        coef_grads, intercept_grads,
+                        sample_weight[batch_slice])
```
`_backprop` is using a subset of the `sample_weights` to weigh the losses. Consider the following sample weights: `[100, 100, 100, 1, 1, 1]` with a batch size is 3, the loss function in `_backprop` will weigh the losses equally.

jnothman replied:
We have considered weighed mini batching, but there was an opinion that this would be okay after enough iterations.

thomasjpfan replied:
If `batch_size=3` and `sample_weight=[100, 100, 100, 1, 1, 1]`, `self._backprop` will be passed a `sample_weight` of `[100, 100, 100]` and `[1, 1, 1]`. When these sample weights are later passed to the loss functions in `self._backprop`, the loss will be weighted uniformly. This will happen in every iteration.

jnothman replied:
So should we adjust the learning rate to account for the overall weight of the mini batch?

dorcoh replied:
Hi, if I may, I think the solution would be to accumulate the loss sample wise instead of batch wise.
If so then some more changes in implementation are needed since loss functions are returning scalars, so first we should change them to output numpy arrays. Then `_backprop` should return `batch_loss` also as array, then `_fit_stochastic` would concatenate the `batch_loss` and finally perform weighted average when each epoch ends.

thomasjpfan replied:
@dorcoh Your idea works, but it will change the loss functions API. Since the loss functions are under `_base.py`, I suspect that are private functions, which means we can change their API without a deprecation cycle. Is this correct @jnothman ?

thomasjpfan reviewed (commented):
There is an issue with backpropagation when calculating losses with sample weights.

jnothman commented:
Yes we can change their API imo

dorcoh committed:
Merge branch 'upstream-master' into mlp_class_weights

# Conflicts:
#	sklearn/metrics/tests/test_common.py
#	sklearn/neural_network/_base.py
#	sklearn/neural_network/multilayer_perceptron.py
#	sklearn/neural_network/tests/test_mlp.py
#	sklearn/utils/testing.py

dorcoh commented:
Hi @jnothman , there's some issue I wanted to consult with you about. 

In order to handle the `train_test_split` and `shuffle` that are called in `_fit_stochastic` and change the order and contents of `X` and `y` I somehow need to tie each sample weight to its corresponding data point, and I'm not sure about how to achieve that. Glad if  you can have a look and let me know what you think

jnothman commented:
שנה טובה @dorcoh, and sorry for the delay. Both train_test_split and shuffle will accept more positional args, so just as you do `train_test_split(X, y, this=that)` you can do  `train_test_split(X, y, sample_weight, this=that)`. Similar for `shuffle`. Not sure if that's what you were unsure about...

dorcoh committed:
Merge branch 'upstream-master' into mlp_class_weights

dorcoh committed:
Merge branch 'upstream-master' into mlp_class_weights

dorcoh commented:
Hi @jnothman , many thanks and Shana Tova to you too :)

I had some time to go over this PR, implemented some parts with conflicts and ran again the class weights test from `estimator_checks.py`. About the issue that @thomasjpfan raised, it shouldn't effect the result of this specific test since batch size is equal to its train set size (50 samples). About my suggestion to fix that issue well it won't work since we should perform gradient step on each batch and it requires to have value for loss (then we cannot compute the loss only at the end of epoch).

In my attempts to pass the test I tried many different hyper parameters in order to figure out whether the data or the model architecture have anything to do with that. As an extreme case I tried to set the model capacity to 1,000,000 neurons with 1 layer, then after 26 iterations the model predicted class '0' 86% of times (original test requires >87%). Do you think we should search the right configuration that pass the test? I also thought about implementing this test in Keras (where class weight is implemented) to understand if its requirements are reasonable. Please share your comments, thanks !

jnothman commented:
(I'd love some help here from another core dev who can be more available
for it...)

thomasjpfan commented:
> (I'd love some help here from another core dev who can be more available
for it...)

Looking at it now.

@dorcoh Are you having trouble with `check_class_weight_classifiers` with the current state of this PR?

thomasjpfan commented:
If you can please do a merge so that the CI can run on azure and the conflicts are resolved.

dorcoh committed:
fix gradient test to pass sample weight

dorcoh committed:
add support for sample and class weights

dorcoh committed:
add sample weight support for loss functions

dorcoh committed:
remove redundant lines

dorcoh commented:
@thomasjpfan yes indeed, the problematic test is `check_class_weight_classifiers`, I pushed some commits, please have a look

dorcoh commented:
Hi, I added a new test that is supposed to check sample/class weights. Please have a look and let me know if it makes sense for you

dorcoh committed:
add sample and class weights tests for MLP

dorcoh committed:
modified weighted loss test; handled flake8 errors; updated docstring

dorcoh committed:
Merge branch 'upstream-master' into mlp_class_weights

# Conflicts:
#	sklearn/neural_network/_base.py
#	sklearn/neural_network/_multilayer_perceptron.py
#	sklearn/utils/estimator_checks.py

dorcoh committed:
add test for loss functions; align with master

thomasjpfan commented:
Sorry for the delayed response, I will look at this after the 0.23 release.

Thank you for updating the PR.

dorcoh committed:
Merge branch 'upstream-master' into mlp_class_weights

# Conflicts:
#	sklearn/neural_network/_base.py
#	sklearn/neural_network/tests/test_mlp.py

cmarmo commented:
Hi @dorcoh , thanks for your work so far. If you are still interested in working on this do you mind fixing conflicts and lint errors? Thanks!

dorcoh committed:
Merge branch 'upstream-master' into mlp_class_weights

# Conflicts:
#	sklearn/neural_network/_multilayer_perceptron.py

dorcoh committed:
correct pep8 violations

dorcoh commented:
Hi @cmarmo, sure, I will try to handle it soon.

nateclind commented:
Forgive me because I'm somewhat new to this, but where is this at in the process? Is this on pause until the rest of the checks are passed? It would be great to see support for this officially happen.

dorcoh commented:
Hey @thomasjpfan @jnothman - do we want to push this PR forward?
Otherwise I guess we can also close it, perhaps there are other contributors who might provide better approaches for implementing this feature :)

thomasjpfan commented:
To continue with this PR, we need to figure out why `check_class_weight_classifiers` is failing with MLP + `class_weights`.

zshu115x commented:
Hi @jnothman @dorcoh @thomasjpfan, I noticed that this PR hasn't been updated for over 2 years and is hard to rebase. That's why I have started new PRs to address the same issue. 

As a prerequisite for adding class weight support, could you please review PR https://github.com/scikit-learn/scikit-learn/pull/25646 to add sample weight support for MLP first?

(I am working on another PR to support class weight https://github.com/scikit-learn/scikit-learn/pull/25326)

adrinjalali commented:
Sadly I think we're not getting this one: xref: https://github.com/scikit-learn/scikit-learn/issues/9113#issuecomment-3334498838

Thank you very much for your work on the issue @dorcoh  and sorry for this work not getting merged.
